src/sparseml/optim/helpers.py

```diff
@@ -512,3 +517,117 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
+            "It seems that some stages in your checkpoint recipe"
+            "contain metadata and some don't."
+            "This is not allowed: either all the "
+            f"stages contain {RECIPE_METADATA_KEY} key or none does."
+        )
+
+    return metadata
+
+
+def _get_recipe_stage_names(container):
+    # Extracts valid stage names from a container.
+    # Valid stage name (key) is the one which corresponds to a value, which is
+    # a dictionary where at least one of the keys contains a string 'modifiers'.
+
+    stage_names = [
+        stage_name
+        for stage_name, stage_dict in container.items()
+        if isinstance(stage_dict, dict)
+        and any([key for key in stage_dict.keys() if "modifiers" in key])
+    ]
+    return stage_names
+
+
+def validate_metadata(metadata: dict, yaml_str: str) -> dict:
+    """
+    Compare the metadata carried over from the recipe (`yaml_str`) with
+    the new, incoming metadata ('metadata').
+
+    If yaml_str is a staged recipe, this function returns a multi-key
+        dict where:
+        -keys are stage names
+        -values are metadata for respective stage name
+    If yaml_str is a normal recipe, this function returns a single-key
+        dict where:
+        -key is RECIPE_METADATA_KEY
+        -value is metadata for the recipe
+
+    If there exists a key in new metadata which also exists in previous metadata,
+    it will overwrite the old metadata. However, it is prohibitive to pass novel keys.
+
+    :param metadata: New metadata
+    :param yaml_str: String representation of the recipe YAML file,
+        (may contain previous metadata)
+    :return: Validated metadata
+    """
+
+    container = load_recipe_yaml_str_no_classes(yaml_str)
+    is_container_staged = check_if_staged_recipe(container)
+
+    checkpoint_metadata = (
+        _extract_metadata_from_staged_recipe(container)
+        if is_container_staged
+        else _extract_metadata_from_recipe(container)
+    )
+
+    if checkpoint_metadata:
+        if metadata:
+            for var_name, var_value in metadata.items():
+                if is_container_staged:
+                    checkpoint_metadata, key_found = _update_staged_recipe_variable(
+                        var_name, var_value, checkpoint_metadata
+                    )
+                else:
+                    checkpoint_metadata, key_found = _update_recipe_variable(
+                        var_name, var_value, checkpoint_metadata
+                    )
+
+                if not key_found:
+                    raise ValueError(
+                        "Attempting to find and overwrite overwrite checkpoint metadata"
+                        f"with metadata key {var_name}, but {var_name}"
+                        "is not present in checkpoint metadata."
+                    )
+
+        return (
+            checkpoint_metadata
+            if is_container_staged
+            else {RECIPE_METADATA_KEY: checkpoint_metadata}
```

Comment:
let's simplify the logic here:

1. checkpoint metadata + new metadata
       if they are equal return either
       otherwise raise warning with any differences logged
2. return the one that exists (expanding as needed for stages)


```pyton

def _check_warn_dict_difference(stage_name: optional, old_metadata, new_metadata):
    pass


if checkpoint_metadata and metadata:
    if staged:
          for stage_name in stage_names:
                 _check_warn_dict_difference(stage_name, container[stage_name][metadata_key], metadata)
   else:
         _check_warn_dict_difference(None, container[metadata_key], metadata)
  if metadata:
      if staged:
           return {stage_name: metadata for stage_name in stage_names}
      else:
           return metadata
elif checkpoint_metadata:
   return checkpoint_metadata
else:
 pass

```